File: server/models/Exam.js
import mongoose from "mongoose";

/** The six question types CheckWise understands, as stored on disk. */
export const QUESTION_TYPES = [
  "multiple-choice",
  "true-false",
  "modified-true-false",
  "identification",
  "fill-in-the-blanks",
  "enumeration",
];

/** Human labels for UI and exported reports. */
export const QUESTION_TYPE_LABELS = {
  "multiple-choice": "Multiple Choice",
  "true-false": "True or False",
  "modified-true-false": "Modified True or False",
  identification: "Identification",
  "fill-in-the-blanks": "Fill in the Blanks",
  enumeration: "Enumeration",
};

export const EXAM_STATUSES = ["draft", "needs-review", "ready"];

/**
 * One question of any of the six supported types.
 *
 * The shape is deliberately a superset rather than six separate schemas: a
 * teacher may change a question's type on the review screen (Phase 4) without
 * the document having to be recreated. Fields not relevant to the current type
 * simply stay empty.
 *
 *   multiple-choice      -> correctAnswers: ["B"],  choices: ["A","B","C","D"]
 *   true-false           -> correctAnswers: ["TRUE"], choices: ["TRUE","FALSE"]
 *   modified-true-false  -> truthValue: "FALSE", correctionAnswers: ["JavaScript"]
 *   identification       -> correctAnswers: ["React","ReactJS","React.js"]
 *   fill-in-the-blanks   -> correctAnswers: ["JavaScript"]
 *   enumeration          -> correctAnswers: ["HTML","CSS","JS"], enumerationCount: 3
 *
 * `correctAnswers` holds *acceptable variations* for written types - any one of
 * them scores the mark - but for enumeration it holds the distinct items the
 * student must list.
 */
const questionSchema = new mongoose.Schema(
  {
    /** Unique across the whole exam - what results and grading key on. */
    questionNumber: { type: Number, required: true, min: 1 },
    /**
     * Exams are written in sections that each restart at 1 ("TEST II" begins
     * again at item 1), so the number printed on the paper is kept separately.
     * The answer sheet prints `sectionNumber` under its `section` heading, so
     * what the student sees matches the questionnaire in their hands.
     */
    section: { type: String, default: "", trim: true, maxlength: 120 },
    sectionNumber: { type: Number, default: null, min: 1 },
    questionType: { type: String, enum: QUESTION_TYPES, required: true },
    questionText: { type: String, default: "", trim: true, maxlength: 2000 },
    correctAnswers: { type: [String], default: [] },
    choices: { type: [String], default: [] },
    truthValue: { type: String, enum: ["TRUE", "FALSE", null], default: null },
    correctionAnswers: { type: [String], default: [] },
    enumerationCount: { type: Number, default: null, min: 1 },
    points: { type: Number, default: 1, min: 0, max: 100 },
    createdAt: { type: Date, default: Date.now },
  },
  { _id: false }
);

const gradingConfigSchema = new mongoose.Schema(
  {
    /**
     * How a Modified True or False question is worth marks.
     *   "whole" - 1 point, and BOTH the truth value and the correction must be right
     *   "split" - truth value 1 point + correction 1 point, scored independently
     */
    modifiedTrueFalseScoring: {
      type: String,
      enum: ["whole", "split"],
      default: "whole",
    },
    /** Enumeration awards one point per correct item rather than all-or-nothing. */
    enumerationPartialCredit: { type: Boolean, default: true },

    /**
     * Written answers must match the key exactly.
     *
     * Strict is the default because the answer key is the teacher's ruling, and
     * software should not quietly widen it. The cost is that the handwriting
     * reader's own misreadings ("usedtate" for "useState") then land on the
     * student, so an exam scored from scans may want this turned off.
     */
    strictWrittenAnswers: { type: Boolean, default: true },
  },
  { _id: false }
);

const examSchema = new mongoose.Schema(
  {
    teacherId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
      index: true,
    },
    title: { type: String, required: true, trim: true, maxlength: 160 },
    subject: { type: String, required: true, trim: true, maxlength: 120 },
    description: { type: String, default: "", trim: true, maxlength: 1000 },
    passingScore: { type: Number, required: true, min: 1, max: 100, default: 75 },

    gradingConfig: { type: gradingConfigSchema, default: () => ({}) },

    // Populated in Phase 3 (upload) and Phase 4 (parse).
    examPdfPath: { type: String, default: null },
    examPdfOriginalName: { type: String, default: null },
    answerKeyPath: { type: String, default: null },
    answerKeyOriginalName: { type: String, default: null },

    questions: { type: [questionSchema], default: [] },
    answerKeyConfirmed: { type: Boolean, default: false },

    examCode: { type: String, required: true, unique: true, index: true },
    answerSheetPath: { type: String, default: null },

    /**
     * Where every bubble sits on the generated sheet, in PDF points.
     *
     * Written by the sheet generator and read by the scanner, so a printed
     * paper is always read with the geometry it was printed with - changing the
     * layout later cannot silently misread sheets already in students' hands.
     * Deliberately schemaless: it is an opaque record produced and consumed by
     * one pair of services, never queried.
     */
    answerSheetLayout: { type: mongoose.Schema.Types.Mixed, default: null },

    /**
     * Denormalised so exam lists and dashboard widgets stay a single indexed
     * read - .lean() queries skip virtuals, so these must be real fields.
     * Recomputed from `questions` on every save.
     */
    totalQuestions: { type: Number, default: 0, min: 0 },
    totalPoints: { type: Number, default: 0, min: 0 },
    status: { type: String, enum: EXAM_STATUSES, default: "draft", index: true },
  },
  { timestamps: true }
);

// Search by title/subject/code from the exam list.
examSchema.index({ teacherId: 1, createdAt: -1 });

/** Derives the counters and workflow status from what is actually stored. */
function syncDerivedFields(doc) {
  const questions = doc.questions || [];
  doc.totalQuestions = questions.length;
  doc.totalPoints = questions.reduce((sum, q) => sum + (q.points ?? 0), 0);

  if (questions.length === 0) {
    doc.status = "draft";
    // A confirmed key cannot survive its questions being removed.
    doc.answerKeyConfirmed = false;
  } else {
    doc.status = doc.answerKeyConfirmed ? "ready" : "needs-review";
  }
}

examSchema.pre("save", function syncOnSave(next) {
  syncDerivedFields(this);
  next();
});

examSchema.virtual("isReadyToScan").get(function isReadyToScan() {
  return this.status === "ready" && this.questions.length > 0;
});

examSchema.set("toJSON", { virtuals: true });

export default mongoose.model("Exam", examSchema);

File: mobile/lib/models/exam.dart
import 'user.dart';

/// The six question types CheckWise understands, and their human labels.
/// Kept in step with `server/models/Exam.js`.
const questionTypeLabels = <String, String>{
  'multiple-choice': 'Multiple Choice',
  'true-false': 'True or False',
  'modified-true-false': 'Modified True or False',
  'identification': 'Identification',
  'fill-in-the-blanks': 'Fill in the Blanks',
  'enumeration': 'Enumeration',
};

/// Bubble types are read by the OMR pass; the rest are handwriting, and are the
/// ones a teacher may still have to type in.
const bubbleQuestionTypes = <String>{
  'multiple-choice',
  'true-false',
  'modified-true-false',
};

String questionTypeLabel(String type) => questionTypeLabels[type] ?? type;

class Question {
  const Question({
    required this.questionNumber,
    required this.questionType,
    this.section = '',
    this.sectionNumber,
    this.questionText = '',
    this.correctAnswers = const [],
    this.choices = const [],
    this.truthValue,
    this.correctionAnswers = const [],
    this.enumerationCount,
    this.points = 1,
  });

  factory Question.fromJson(Map<String, dynamic> json) {
    return Question(
      questionNumber: asInt(json['questionNumber']),
      questionType: asString(json['questionType']),
      section: asString(json['section']),
      sectionNumber: json['sectionNumber'] == null
          ? null
          : asInt(json['sectionNumber']),
      questionText: asString(json['questionText']),
      correctAnswers: asStringList(json['correctAnswers']),
      choices: asStringList(json['choices']),
      truthValue: json['truthValue'] == null
          ? null
          : asString(json['truthValue']),
      correctionAnswers: asStringList(json['correctionAnswers']),
      enumerationCount: json['enumerationCount'] == null
          ? null
          : asInt(json['enumerationCount']),
      points: asDouble(json['points'], 1),
    );
  }

  final int questionNumber;
  final String questionType;
  final String section;
  final int? sectionNumber;
  final String questionText;

  /// Acceptable variations for written types; the distinct items for
  /// enumeration.
  final List<String> correctAnswers;
  final List<String> choices;
  final String? truthValue;
  final List<String> correctionAnswers;
  final int? enumerationCount;
  final double points;

  /// What the student sees printed on the paper, which restarts per section.
  int get displayNumber => sectionNumber ?? questionNumber;

  bool get isBubbled => bubbleQuestionTypes.contains(questionType);
}

/// One page of an exam list, matching the API's `pagination` block.
class Pagination {
  const Pagination({
    this.page = 1,
    this.limit = 10,
    this.total = 0,
    this.totalPages = 1,
    this.hasPrev = false,
    this.hasNext = false,
  });

  factory Pagination.fromJson(Map<String, dynamic> json) {
    return Pagination(
      page: asInt(json['page'], 1),
      limit: asInt(json['limit'], 10),
      total: asInt(json['total']),
      totalPages: asInt(json['totalPages'], 1),
      hasPrev: asBool(json['hasPrev']),
      hasNext: asBool(json['hasNext']),
    );
  }

  final int page;
  final int limit;
  final int total;
  final int totalPages;
  final bool hasPrev;
  final bool hasNext;
}

class ExamPage {
  const ExamPage({required this.exams, required this.pagination});

  factory ExamPage.fromJson(Map<String, dynamic> json) {
    final list = json['exams'];
    return ExamPage(
      exams: list is List
          ? list
                .whereType<Map<String, dynamic>>()
                .map(Exam.fromJson)
                .toList()
          : const [],
      pagination: Pagination.fromJson(
        (json['pagination'] as Map<String, dynamic>?) ?? const {},
      ),
    );
  }

  final List<Exam> exams;
  final Pagination pagination;
}

class Exam {
  const Exam({
    required this.id,
    required this.title,
    required this.subject,
    required this.examCode,
    required this.status,
    this.description = '',
    this.passingScore = 75,
    this.totalQuestions = 0,
    this.totalPoints = 0,
    this.answerKeyConfirmed = false,
    this.hasAnswerSheet = false,
    this.hasExamPdf = false,
    this.strictWrittenAnswers = true,
    this.modifiedTrueFalseScoring = 'whole',
    this.questions = const [],
    this.createdAt,
    this.updatedAt,
  });

  factory Exam.fromJson(Map<String, dynamic> json) {
    final grading = (json['gradingConfig'] as Map<String, dynamic>?) ?? const {};
    final questions = json['questions'];

    return Exam(
      id: asId(json),
      title: asString(json['title']),
      subject: asString(json['subject']),
      examCode: asString(json['examCode']),
      status: asString(json['status'], 'draft'),
      description: asString(json['description']),
      passingScore: asInt(json['passingScore'], 75),
      totalQuestions: asInt(json['totalQuestions']),
      totalPoints: asDouble(json['totalPoints']),
      answerKeyConfirmed: asBool(json['answerKeyConfirmed']),
      // The layout is written only when the printable sheet has been built,
      // and the scanner cannot read a paper without it.
      hasAnswerSheet: json['answerSheetLayout'] != null ||
          json['answerSheetPath'] != null,
      hasExamPdf: json['examPdfPath'] != null,
      strictWrittenAnswers: asBool(grading['strictWrittenAnswers'], true),
      modifiedTrueFalseScoring:
          asString(grading['modifiedTrueFalseScoring'], 'whole'),
      // Absent from list rows, which select the questions array away.
      questions: questions is List
          ? questions
                .whereType<Map<String, dynamic>>()
                .map(Question.fromJson)
                .toList()
          : const [],
      createdAt: asDate(json['createdAt']),
      updatedAt: asDate(json['updatedAt']),
    );
  }

  final String id;
  final String title;
  final String subject;
  final String examCode;
  final String status;
  final String description;
  final int passingScore;
  final int totalQuestions;
  final double totalPoints;
  final bool answerKeyConfirmed;
  final bool hasAnswerSheet;
  final bool hasExamPdf;
  final bool strictWrittenAnswers;

  /// "whole" — one point, and both halves must be right.
  /// "split" — the truth value and the correction score independently.
  final String modifiedTrueFalseScoring;
  final List<Question> questions;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  String get statusLabel => switch (status) {
    'ready' => 'Ready',
    'needs-review' => 'Needs review',
    _ => 'Draft',
  };

  /// A paper can only be scored against a confirmed key, and only read with the
  /// geometry the sheet was printed with.
  bool get canScan => status == 'ready' && hasAnswerSheet;

  /// Why scanning is unavailable, phrased for the teacher. Both blockers are
  /// fixable on the phone now, so these name the step rather than send them
  /// somewhere else.
  String? get scanBlockedReason {
    if (status != 'ready') {
      return 'Confirm this exam’s answer key before scanning papers.';
    }
    if (!hasAnswerSheet) {
      return 'Generate this exam’s answer sheet first — the scanner reads its '
          'layout to know where the bubbles are.';
    }
    return null;
  }

  /// Sections in printed order, for grouping the review screen.
  List<String> get sections {
    final seen = <String>[];
    for (final question in questions) {
      if (question.section.isNotEmpty && !seen.contains(question.section)) {
        seen.add(question.section);
      }
    }
    return seen;
  }
}
